src/coreclr/jit/helperexpansion.cpp

```diff
@@ -1171,3 +1178,331 @@ bool Compiler::fgExpandStaticInitForCall(BasicBlock** pBlock, Statement* stmt, G
+//
+// Returns:
+//    True if expanded, false otherwise.
+//
+bool Compiler::fgVNBasedIntrinsicExpansionForCall_ReadUtf8(BasicBlock** pBlock, Statement* stmt, GenTreeCall* call)
+{
+    BasicBlock* block = *pBlock;
+
+    assert(call->gtArgs.CountUserArgs() == 4);
+
+    GenTree* srcPtr = call->gtArgs.GetUserArgByIndex(0)->GetNode();
+
+    // We're interested in a case when srcPtr is a string literal and srcLen is a constant
+    // srcLen doesn't have to match the srcPtr's length, but it should not exceed it.
+    ssize_t               strObjOffset = 0;
+    CORINFO_OBJECT_HANDLE strObj       = nullptr;
+    if (!GetObjectHandleAndOffset(srcPtr, &strObjOffset, &strObj) || ((size_t)strObjOffset > INT_MAX))
+    {
+        // We might want to support more cases here, e.g. ROS<char> RVA data.
+        // Also, we check that strObjOffset (which is in most cases is expected to be just
+        // OFFSETOF__CORINFO_String__chars) doesn't exceed INT_MAX since we'll need to cast
+        // it to int for getObjectContent API.
+        JITDUMP("ReadUtf8: srcPtr is not an object handle\n")
+        return false;
+    }
+
+    // We mostly expect string literal objects here, but let's be more agile just in case
+    if (!info.compCompHnd->isObjectImmutable(strObj))
+    {
+        JITDUMP("ReadUtf8: srcPtr is not immutable (not a frozen string object?)\n")
+        return false;
+    }
+
+    GenTree* srcLen = call->gtArgs.GetUserArgByIndex(1)->GetNode();
+    if (!srcLen->gtVNPair.BothEqual() || !vnStore->IsVNInt32Constant(srcLen->gtVNPair.GetLiberal()))
+    {
+        JITDUMP("ReadUtf8: srcLen is not constant\n")
+        return false;
+    }
+
+    const int      MaxPossibleUnrollThreshold = 256;
+    const unsigned unrollThreshold            = min(getUnrollThreshold(UnrollKind::Memcpy), MaxPossibleUnrollThreshold);
+    const unsigned srcLenCns                  = (unsigned)vnStore->GetConstantInt32(srcLen->gtVNPair.GetLiberal());
+    if ((srcLenCns == 0) || (srcLenCns > unrollThreshold))
+    {
+        // TODO: handle srcLenCns == 0 if it's a common case
+        JITDUMP("ReadUtf8: srcLenCns is out of unrollable range\n")
+        return false;
+    }
+
+    // Read the string literal (UTF16) into a local buffer (UTF8)
+    assert(strObj != nullptr);
+    BYTE buffer[MaxPossibleUnrollThreshold * 2];
+
+    // Both must be within [0..INT_MAX] range as we're going to cast them to int
+    assert((unsigned)srcLenCns <= INT_MAX);
+    assert((unsigned)strObjOffset <= INT_MAX);
+
+    // getObjectContent is expected to validate the offset and length
+    if (!info.compCompHnd->getObjectContent(strObj, buffer, (int)srcLenCns * 2, (int)strObjOffset))
+    {
+        JITDUMP("ReadUtf8: getObjectContent returned false.\n")
+        return false;
+    }
+
+    for (unsigned charIndex = 0; charIndex < srcLenCns; charIndex++)
+    {
+        // Buffer keeps the original utf16 chars
+        uint16_t ch = ((uint16_t*)buffer)[charIndex];
```

Comment:
Interpreting this buffer as both `BYTE` and `uint16_t` with pointer casts makes me feel a bit uneasy... maybe `memcpy` this?

Reply:
Thanks, decided to introduce a separate buffer for output - 256 bytes won't hurt it